Add PostgreSQL Decision Page (#32)
vshn/application-catalog-docs: docs/modules/ROOT/pages/explanations/decisions/postgresql.adoc

= PostgreSQL Operators

== Problem

Creating our own operator to deploy and manage PostgreSQL proved to be a gargantuan effort.

> In order to keep in line with the newly specified https://kb.vshn.ch/app-catalog/reference/framework-requirements.html#_for_service_maintainer["more buy than make"] approach, we're evaluating the viability of using existing operators for the use of converged services.

Reviewer: ```suggestion
In order to keep in line with the newly specified xref:reference/framework-requirements.adoc#_for_service_maintainer["more buy than make"] approach, we're evaluating the viability of using existing operators for the use of converged services.
```

= PostgreSQL Operators

== Problem

> Creating our own operator to deploy and manage PostgreSQL proved to be a gargantuan effort. In order to keep in line with the newly specified https://kb.vshn.ch/app-catalog/reference/framework-requirements.html#_for_service_maintainer["more buy than make"] approach, we're evaluating the viability of using existing operators for the use of converged services.

Reviewer: Maybe state a few more issues we've seen, besides the effort:
* Why build on our own when there are already plenty of viable alternatives around?
* Unpredictable amount of maintenance needed to keep our own solution up-to-date and maintained
* What do we gain when building our own vs. using an existing solution and integrating it into our systems

== Evaluated Operators

[cols="1,1,1,1,1,1,1"]
|===

> |Requirements |Percona |Stackgres |CloudnativePG |Zalando |Crunchydata |Bitnami Helmchart

Reviewer: Link headings to the corresponding websites

|Superuser Access |✅ |✅ |✅ |✅ |✅ |✅

|Database Management |❌ |❌ |❌ |✅ |✅ |❌

|User Management |❌ |❌ |❌ |✅ |✅ |✅

|Service Metrics |❌ |✅ |✅ |❌ |✅ |✅

|Maintenance Schedule |✅ |✅ |❌ |❌ |❌ |❌

|Backup Schedule |✅  |✅  |✅  |✅  |✅  |❌

|Self Service Backup Restore |✅ |✅ |✅ |✅ |✅ |❌

|Encryption at rest |✅ |✅ |✅ |✅ |✅ |✅

|Connection limit of 25/Gb Memory |✅ |✅ |✅ |❌ |✅ |✅

|Extension Management |❌ |✅ |❌ |✅ |❌ |✅

|Custom PostgreSQL Settings |✅ |✅ |✅ |❌ |✅ |✅

|In-Place Major Upgrade |✅ |✅ |❌ |❌ |❌ |❌

|License |Apache License 2.0 |AGPL-3.0 |Apache License 2.0 |MIT |Apache License 2.0 | Apache License 2.0

|Upstream Support |Some support available not clear if for operators |✅ |✅ (via EDB) |❌ |✅ |❌
|===

> https://wiki.vshn.net/display/VHT/PostgreSQL+Operator+Evaluations[More detailed] comparison is in our internal wiki.

Reviewer: Please do not link to our wiki, that's not useful for external readers. I'd extract the most important parts of the wiki page and when this page is released, delete the wiki page. Otherwise, I'm sure this would cause friction.

Reviewer 2: I also favor to integrate the wiki page here.

> For the POC on which this decision is based on, we opted to test out Stackgres. It's feature set and general feel were good during the evaluation phase.

Reviewer: ```suggestion
For the POC on which this decision is based on, we opted to test out Stackgres.
It's feature set and general feel were good during the evaluation phase.
```

Each sentence belongs to its own line.

== Pros

* Brings cluster functionality

> * Backups out of the box, incl. PIT

Reviewer: PIT?

Author: Point-In-Time

> * Backups out of the box, incl. PIT

Reviewer: ```suggestion
* Backups out of the box, incl. Point-In-Time
```

> * Auto deploys Prometheus service monitors

Reviewer 2: ```suggestion
* Automatic deployments of Prometheus service monitors
```

> * Brings Dbops CRs that handle some operations in a k8s native way

Reviewer: ```suggestion
* Brings DbOps CRs that handle some operations in a k8s native way
```

* Security update schedules can be implemented
* The operator is maintained by a third party
* Commercial support available

> 

Reviewer 2: ```suggestion
* Upgrades to the operator affect all users at the same time, independent of their maintenance window settings.

```

Author: At least for stackgres this statement is not true. It won't restart anything until explicitly done via a DbOps CR.

Reviewer 2: I wasn't referring to DB restarts, but in general if there are breaking changes (either in API or behavior). I guess this isn't only applicable to Stackgres, but for all operators.

Author: Ah then I suggest to rephrase it:

`Upgrading the operator could inntroduce breaking changes`

To some degree we can handle those changes in the compositions. But a good point we should be aware about.

== Cons

* We can't support PostgreSQL version or extensions that the operator doesn't provide
* Crossplane's limitations, as discussed https://kb.vshn.ch/app-catalog/explanations/decisions/converged-service-impl.html[here]

> * We have to deal with some opinionated decision by the operator developers and maybe need to work around them

Reviewer: Or accept them as a given

> * The operator is maintained by a third party

Reviewer: But it's Open Source and we could contribute if needed

== Risks

=== Operator Unmaintained

Risk:: The Operator becomes unmaintained

Option 1:: We or the community carries the project further
Option 2:: Migration to another operator

=== CRD Versioning

Risk:: CRDs could contain breaking changes

> Option 1:: The maintainer provides conversion webhooks

Reviewer 2: I wouldn't consider this as an option, because conversion webhooks do the magic for us and there'd be nothing to do. Or did I miss something?

Author: I've included this, because it mitigates the risk. It's by far the best mitigation for this risk, but maybe not always available.

Reviewer 2: Sure it mitigates it, but this option is not in our hands. I understand these risk assessment as "what are _our_ action options if it occurs".

Maybe we need to rephrase the risk to `CRDs could contain breaking changes that require a manual upgrade path`

Option 2:: We can handle the changes in the compositions, transparent for the user
Option 3:: We inform the customers they have to do an action after the upgrade

=== License Changes

> Risk:: The licensing of the operator could change

Reviewer 2: Changing the licensing is not the problem itself, resp. it matters which kind of license it's being changed to.
```suggestion
Risk:: The licensing of the operator could change to a hostile license
```

Option 1:: Migrate to another operator

> Option 2:: Keep using a version before the change applied

Reviewer 2: Option 2 effectively means that the service is not really supported anymore and it "stalls". Eventually something breaks or newer versions of Postgres don't run anymore.
I would not consider this an option for long-term, rather combine it with option 1, at which point I would just list option 1.
